test/e2e/rebalance_test.go

```diff
@@ -0,0 +1,214 @@
+	r := rebalancer.NewRebalancer(ctx, f.KubeClient)
+	workload := &rebalancer.Workload{Namespace: f.Ns, Name: "prometheus-k8s", LabelSelector: map[string]string{"app.kubernetes.io/name": "prometheus"}}
+
+	// Enable persistent storage
+	err := enablePersistentStorage()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer func() {
+		err := disablePersistentStorage()
+		if err != nil {
+			t.Fatal(err)
+		}
+	}()
+
+	// Rebalance all prometheus-k8s pods on the same node to setup the scenario
+	// where pods needs to be rebalanced by the operator
+	err = incorrectlyRebalanceWorkload(ctx, r, workload)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Annotate prometheus-k8s-0 PVC for deletion
+	err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		pvc, err := f.KubeClient.CoreV1().PersistentVolumeClaims(f.Ns).Get(ctx, "prometheus-k8s-db-prometheus-k8s-0", metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		pvc.Annotations[rebalancer.DropPVCAnnotation] = "yes"
+		_, err = f.KubeClient.CoreV1().PersistentVolumeClaims(f.Ns).Update(ctx, pvc, metav1.UpdateOptions{})
+		return err
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Wait for the automated rebalancing to quick in
+	err = framework.Poll(time.Second, 5*time.Minute, func() error {
+		balanced, err := r.WorkloadCorrectlyBalanced(ctx, workload)
+		if err != nil {
+			return err
+		}
+		if balanced {
+			return nil
+		}
+		return fmt.Errorf("Expected workload %s/%s to be correctly balanced between multiple nodes", workload.Namespace, workload.Name)
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Verify that the PVC isn't annotated anymore
+	pvc, err := f.KubeClient.CoreV1().PersistentVolumeClaims(f.Ns).Get(ctx, "prometheus-k8s-db-prometheus-k8s-0", metav1.GetOptions{})
+	if _, found := pvc.Annotations[rebalancer.DropPVCAnnotation]; found {
+		t.Fatalf("Expected %s annotation to have been removed from the PVC %s/%s", rebalancer.DropPVCAnnotation, f.Ns, "prometheus-k8s-db-prometheus-k8s-0")
+	}
+
+	// Verify that no node was left unschedulable by CMO
+	nodeList, err := f.KubeClient.CoreV1().Nodes().List(ctx, metav1.ListOptions{LabelSelector: "!node-role.kubernetes.io/master"})
+	if err != nil {
+		t.Fatal(err)
+	}
+	for _, node := range nodeList.Items {
+		_, cordonedByOperator := node.Annotations[rebalancer.CordonAnnotation]
+		if node.Spec.Unschedulable && cordonedByOperator {
+			t.Fatalf("Expected node %s cordoned by CMO to have been uncordoned after rebalancing the pods", node.Name)
+		}
+	}
+}
+
+func incorrectlyRebalanceWorkload(ctx context.Context, r *rebalancer.Rebalancer, workload *rebalancer.Workload) error {
+	// Cordon all nodes except one
+	nodeList, err := f.KubeClient.CoreV1().Nodes().List(ctx, metav1.ListOptions{LabelSelector: "!node-role.kubernetes.io/master"})
+	if err != nil {
+		return err
+	}
+	drainer := &drain.Helper{Ctx: ctx, Client: f.KubeClient}
+	for _, node := range nodeList.Items[1:] {
+		err = drain.RunCordonOrUncordon(drainer, &node, cmodrain.Cordon)
```

Comment:
by using the same annotation, isn't there a risk of race (e.g. CMO might uncordon the nodes if it runs the sync() loop)?

Reply:
in the e2e tests I am using the `kubectl` cordon for that exact reason

Reply:
:man_facepalming: sorry about the noise!